File: pycodes/LinkList.py
class ListNode:

    def __init__(self, val=0):
        self.val = val
        self.next = None


class MyLinkedList(object):
    
    def __init__(self):
        self.size = 0
        self.head = None

    def get(self, index):
        node = self.getnode(index=index)
        return -1 if node == None else node.val
        
    def addAtHead(self, val):
        self.addAtIndex(0,val=val)


    def addAtTail(self, val):
        self.addAtIndex(self.size,val=val)


    def addAtIndex(self, index, val):
        if index<=self.size:
            add = ListNode(val=val)
            if index == 0:
                add.next = self.head
                self.head = add
            else:
                prev = self.getnode(index-1)
                if prev != None:
                    add.next = prev.next
                    prev.next = add
            self.size += 1
        


    def deleteAtIndex(self, index):
        if index < self.size:
            if index == 0:
                self.head = self.head.next
            else:
                prev = self.getnode(index-1)
                node = prev.next
                prev.next = node.next
            self.size -= 1
    
    def getnode(self,index):
        temp = self.head
        while index > 0:
            if temp.next == None:
                return None
            index -= 1
            temp = temp.next
        return temp
    
class Solution():
    def _203(self, head, val):
        if head == None: return
        while True:
            if head.val==val:
                if head.next ==None:
                    return
                head = head.next
            else: break
        a, b = head, head.next
        while b != None:
            if b.val == val:
                b = b.next
                a.next = b
            else:
                a = a.next
                b = b.next
        return head
    
    def _206(head):
        if head == None: return
        a = head.next
        if a == None: return
        head.next = None
        while a != None:
            print(1)
            temp = a.next
            a.next = head
            head = a
            a = temp
        return head

    def _24(self, head):
        if head == None: return None
        a,b = head, head.next
        if b == None:return a
        while True:
            if a == None or b == None:break
            a.val, b.val = b.val, a.val
            if b.next == None:break
            a = b.next
            b = a.next
        return head

    def _19(self, head, n):
        dummyhead = ListNode()
        dummyhead.next = head
        fast = slow = dummyhead
        for i in range(n):
            fast = fast.next
        while fast.next != None:
            slow = slow.next
            fast = fast.next
        tmp = slow.next
        slow.next = tmp.next
        return dummyhead.next

    def getIntersectionNode(self, headA, headB):
        # 链表相交的题目，这个思路很巧妙。！！！
        A, B = headA, headB
        while A != B:
            A = A.next if A else headB
            B = B.next if B else headA
        return A
    
    def _142_1(self, head):
        nodelist = []
        while head != None:
            if head in nodelist: return head
            nodelist.append(head)
            head = head.next
        return None

a = MyLinkedList()
a.addAtTail(val=7)
print(a.get(1))

File: cpp/code/linklist.cpp
#include<iostream>
#include<limits>
#include<math.h>
#include<vector>

using namespace std;


// // 定义链表节点
// struct ListNode{
//     int val; // 节点的值
//     ListNode *next; // 指向下一个节点
//     /*  节点的构造函数，cpp有一个默认的构造函数，默认是无参数的构造，不能传入参数
//         但是可以像下面一样自定义一个构造函数，但是必需按照构造来传入参数，不然报错    
//     */
//     ListNode(int x): val(x), next(NULL) {}
// };


// *
//  * Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

// 707 设计链表
class MyLinkedList {
public:

    MyLinkedList() {
        dummyhead = new ListNode();
        _size = 0;
    }
    
    int get(int index) {
        if(index >= _size or index < 0){
            return -1;
        }
        ListNode *res = dummyhead;
        while(index >= 0){
            res = res->next;
            index--;
        }
        return res->val;
    }
    
    void addAtHead(int val) {
        ListNode* node = new ListNode();
        node->val = val;
        node->next = dummyhead->next;
        dummyhead->next = node;
        _size++;
    }
    
    void addAtTail(int val) {
        ListNode* cur = dummyhead;
        while(cur->next){
            cur = cur->next;
        }
        ListNode* node = new ListNode;
        node->val = val;
        cur->next = node;
        _size++;
    }
    
    void addAtIndex(int index, int val) {
        if(index >= 0 and index <= _size){
            ListNode* cur = dummyhead;
            while(index > 0){
                cur = cur->next;
                index--;
            }
            ListNode* node = new ListNode();
            node->val = val;
            node->next = cur->next;
            cur->next = node;
            _size++;
        }
    }
    
    void deleteAtIndex(int index) {
        if(index >= 0 and index < _size){
            ListNode* cur = dummyhead;
            while(index > 0){
                cur = cur->next;
                index--;
            }
            ListNode* tmp = cur->next;
            cur->next = tmp->next;
            delete tmp;
            _size--;
        }
    }

private:
    ListNode* dummyhead;
    int _size;
};

class Solution {
public:

    // 203 移除链表元素
    // 我的方法，设置一个dummyhead
    ListNode* removeElements(ListNode* head, int val) {
        ListNode* vhead = new ListNode(0, head);
        return deleteElementss(vhead, val);
    }
    ListNode* deleteElementss(ListNode* head, int val){
        ListNode* res = head;
        while ( head->next != nullptr){
            if (head->next->val == val){
                ListNode *tmp = head->next;
                head->next = head->next->next;
                delete tmp;
            }
            else{
                head = head->next;
            }
        }
        return res->next;
    }

    // 参考
    ListNode* removeElements_ref(ListNode* head, int val) {
        ListNode* dummyhead = new ListNode();
        dummyhead->next = head;
        ListNode* cur = dummyhead;
        while(cur->next){
            if(cur->next->val == val){
                ListNode* tmp = cur->next;
                cur->next = tmp->next;
                delete tmp;
            }
            else{
                cur = cur->next;
            }
        }
        head = dummyhead->next;
        delete dummyhead;
        return head;
    }

    // 206 反转链表
    ListNode* reverseList(ListNode* head) {
        ListNode *prev = NULL, *cur = head;
        ListNode *nextnode = NULL;
        while(cur){
            nextnode = cur->next;
            cur->next = prev;
            prev = cur;
            cur = nextnode;
        }
        return prev;
    }

    // 24. 两两交换链表中的节点
    // 自己写的好乱啊,cpp不熟悉,导致逻辑也非常混乱.
    ListNode* swapPairs(ListNode* head) {
        if(!head) return head;
        if(!head->next) return head;
        ListNode *dommyhead = new ListNode();
        ListNode *left = head, *right = head->next;
        ListNode *pre = dommyhead;
        while(true){
            pre->next = right;
            left->next = right->next;
            right->next=left;
            pre = left;
            if(left->next and left->next->next){
                left = left->next;
                right = left->next;
            }
            else break;
        }
        return dommyhead->next;
    }

    // 19. 删除链表的倒数第 N 个结点
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *dummyhead = new ListNode();
        ListNode *pre = dummyhead;
        ListNode *cur = dummyhead;
        dummyhead->next = head;
        while(n > 0){
            cur = cur->next;
            n--;
        }
        while (cur->next){
            pre = pre->next;
            cur = cur->next;
        }
        ListNode *tmp = pre->next;
        pre->next = tmp->next;
        delete tmp;
        return dummyhead->next;
    }

    // 面试题 02.07. 链表相交

    int getListLen(ListNode *head){
        int len = 0;
        while(head){
            len++;
            head = head->next;
        }
        return len;
    }
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        int lenA = getListLen(headA), lenB = getListLen(headB);
        if(lenA < lenB){
            swap(lenA, lenB);
            swap(headA, headB);
        }
        int gap = lenA - lenB;
        while(gap){
            headA = headA->next;
            gap--;
        }
        while(headA){
            if(headA == headB) return headA;
            headA = headA->next;
            headB = headB->next;
        }
        return NULL;
    }

    // 142 环形链表Ⅱ
    // 注意这个问题的数学推导
    ListNode *detectCycle(ListNode *head) {
        ListNode* fast = head, *slow = head;
        while(fast and fast->next){
            fast = fast->next->next;
            slow = slow->next;
            if(slow == fast){
                ListNode* index1 = fast;
                ListNode* index2 = head;
                while(index1 != index2){
                    index1 = index1->next;
                    index2 = index2->next;
                }
                return index1;
            }
        }
        return NULL;
    }
    // 242. 有效的字母异位词
    bool isAnagram(string s, string t) {
        int table[26];
        for(char ch:s){
            table[ch - 'a']++;
        }
        for(char ch:s){
            table[ch - 'a']--;
        }
        for(int times: table){
            if(times) return false;
        }
        return true;
    }

    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
        
    }

};



int main(){
}